File: Java/longestSubstringWithoutRepetition.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        if(s.length()==0)
            return 0;
        int ans= 1;

        char[] c= s.toCharArray();
        int n= c.length;
        int si=0, ei= 0;        

        HashSet<Character> check= new HashSet<>();

        while(ei<n && si<=ei){
            if(!check.contains(c[ei])){
                check.add(c[ei]);
                ei+=1;
            }else{
                ans= Math.max(ans, check.size());
                check.remove(c[si]);
                si+=1;
            }
        }

             
        return Math.max(ans, check.size());
    }
}

File: Python/longestSubstringWithoutRepetition.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        if len(s) == 0:
            return 0
        ans = 1
        n = len(s)
        si, ei = 0, 0
        check = set()

        while ei < n and si <= ei:
            if s[ei] not in check:
                check.add(s[ei])
                ei += 1
            else:
                ans = max(ans, len(check))
                check.remove(s[si])
                si += 1

        return max(ans, len(check))
